Build fails with strict-aliasing violations

I tried to build with the following *FLAGS to optimize the build: `-flto=4 -Werror=odr -Werror=lto-type-mismatch -Werror=strict-aliasing`

The -Werror=* flags are important to detect cases where the compiler can try to optimize based on assuming UB cannot happen, and miscompile code that has UB in it. strict-aliasing issues are always bad but LTO can make them even worse.

I got this error:
```
[1/6] /usr/bin/x86_64-pc-linux-gnu-g++ -DBOOST_CHRONO_HEADER_ONLY -DBOOST_CONFIG_SUPPRESS_OUTDATED_MESSAGE -DNOVA_SIMD -DReverbUGens_EXPORTS -DSC_DATA_DIR=\"/usr/share/SuperCollider\" -DSC_PLUGIN_EXT=\".so\" -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999_build/common -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/external_libraries -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/external_libraries/nova-simd -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/external_libraries/nova-tt -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/include/common -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/common -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/include/plugin_interface  -march=native -fstack-protector-all -O2 -pipe -fdiagnostics-color=always -frecord-gcc-switches -flto=4 -Werror=odr -Werror=lto-type-mismatch -Werror=strict-aliasing  -Wformat -Werror=format-security -DNDEBUG -msse -mfpmath=sse -msse2 -D_LARGEFILE64_SOURCE -fvisibility-inlines-hidden -std=gnu++17 -fPIC   -fschedule-insns2 -fomit-frame-pointer -Wreturn-type -fno-math-errno -fno-signaling-nans -fsigned-zeros -fno-associative-math -pthread -fvisibility=hidden -fno-finite-math-only -MD -MT server/plugins/CMakeFiles/ReverbUGens.dir/ReverbUGens.cpp.o -MF server/plugins/CMakeFiles/ReverbUGens.dir/ReverbUGens.cpp.o.d -o server/plugins/CMakeFiles/ReverbUGens.dir/ReverbUGens.cpp.o -c /var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/server/plugins/ReverbUGens.cpp
FAILED: server/plugins/CMakeFiles/ReverbUGens.dir/ReverbUGens.cpp.o 
/usr/bin/x86_64-pc-linux-gnu-g++ -DBOOST_CHRONO_HEADER_ONLY -DBOOST_CONFIG_SUPPRESS_OUTDATED_MESSAGE -DNOVA_SIMD -DReverbUGens_EXPORTS -DSC_DATA_DIR=\"/usr/share/SuperCollider\" -DSC_PLUGIN_EXT=\".so\" -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999_build/common -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/external_libraries -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/external_libraries/nova-simd -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/external_libraries/nova-tt -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/include/common -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/common -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/include/plugin_interface  -march=native -fstack-protector-all -O2 -pipe -fdiagnostics-color=always -frecord-gcc-switches -flto=4 -Werror=odr -Werror=lto-type-mismatch -Werror=strict-aliasing  -Wformat -Werror=format-security -DNDEBUG -msse -mfpmath=sse -msse2 -D_LARGEFILE64_SOURCE -fvisibility-inlines-hidden -std=gnu++17 -fPIC   -fschedule-insns2 -fomit-frame-pointer -Wreturn-type -fno-math-errno -fno-signaling-nans -fsigned-zeros -fno-associative-math -pthread -fvisibility=hidden -fno-finite-math-only -MD -MT server/plugins/CMakeFiles/ReverbUGens.dir/ReverbUGens.cpp.o -MF server/plugins/CMakeFiles/ReverbUGens.dir/ReverbUGens.cpp.o.d -o server/plugins/CMakeFiles/ReverbUGens.dir/ReverbUGens.cpp.o -c /var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/server/plugins/ReverbUGens.cpp
In file included from /var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/include/plugin_interface/SC_PlugIn.h:25,
                 from /var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/server/plugins/ReverbUGens.cpp:18:
/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/server/plugins/ReverbUGens.cpp: In function ‘void gverb_set_roomsize(GVerb*, float)’:
/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/include/plugin_interface/SC_Unit.h:95:82: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
   95 | #define CALCSLOPE(next, prev) ((next - prev) * sc_typeof_cast(next) unit->mRate->mSlopeFactor)
      |                                                                     ~~~~~~~~~~~~~^~~~~~~~~~~~
/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/server/plugins/ReverbUGens.cpp:1201:34: note: in expansion of macro ‘CALCSLOPE’
 1201 |         unit->fdngainslopes[i] = CALCSLOPE(unit->fdngains[i], oldfdngain);
      |                                  ^~~~~~~~~
/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/include/plugin_interface/SC_Unit.h:95:82: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
   95 | #define CALCSLOPE(next, prev) ((next - prev) * sc_typeof_cast(next) unit->mRate->mSlopeFactor)
      |                                                                     ~~~~~~~~~~~~~^~~~~~~~~~~~
/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/server/plugins/ReverbUGens.cpp:1212:34: note: in expansion of macro ‘CALCSLOPE’
 1212 |         unit->tapgainslopes[i] = CALCSLOPE(unit->tapgains[i], oldtapgain);
      |                                  ^~~~~~~~~
/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/server/plugins/ReverbUGens.cpp: In function ‘void gverb_set_revtime(GVerb*, float)’:
/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/include/plugin_interface/SC_Unit.h:95:82: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
   95 | #define CALCSLOPE(next, prev) ((next - prev) * sc_typeof_cast(next) unit->mRate->mSlopeFactor)
      |                                                                     ~~~~~~~~~~~~~^~~~~~~~~~~~
/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/server/plugins/ReverbUGens.cpp:1230:34: note: in expansion of macro ‘CALCSLOPE’
 1230 |         unit->fdngainslopes[i] = CALCSLOPE(unit->fdngains[i], oldfdngain);
      |                                  ^~~~~~~~~
cc1plus: some warnings being treated as errors
[2/6] /usr/bin/x86_64-pc-linux-gnu-g++ -DBOOST_CHRONO_HEADER_ONLY -DBOOST_CONFIG_SUPPRESS_OUTDATED_MESSAGE -DHAVE_ALSA=1 -DHAVE_HIDAPI -DHAVE_LID -DHAVE_READLINE -DNO_INTERNAL_SERVER -DQT_COLLIDER_EXPORTING -DQT_CORE_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_NO_DEBUG -DQT_NO_KEYWORDS -DQT_PRINTSUPPORT_LIB -DQT_SVG_LIB -DQT_WIDGETS_LIB -DSC_AUDIO_API_PORTAUDIO -DSC_DATA_DIR=\"/usr/share/SuperCollider\" -DSC_HIDAPI -DSC_IDE -DSC_QT -DSNDFILE_HAS_MPEG -DSNDFILE_HAS_OPUS -DSNDFILE_HAS_VORBIS -DYYSTACK_USE_ALLOC -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999_build/common -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/include/common -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/include/lang -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/include/plugin_interface -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/include/server -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/common -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/external_libraries/boost_sync/include -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangPrimSource -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/external_libraries/nova-tt -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource/Bison -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/external_libraries/hidapi/hidapi -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/external_libraries/hidapi/hidapi_parser -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/QtCollider -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/X11_INCLUDE_DIR -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/SCDoc -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/external_libraries/TLSF-2.4.6/src -isystem /usr/include/qt5 -isystem /usr/include/qt5/QtCore -isystem /usr/lib64/qt5/mkspecs/linux-g++ -isystem /usr/include/qt5/QtGui -isystem /usr/include/qt5/QtNetwork -isystem /usr/include/qt5/QtPrintSupport -isystem /usr/include/qt5/QtWidgets -isystem /usr/include/qt5/QtSvg  -march=native -fstack-protector-all -O2 -pipe -fdiagnostics-color=always -frecord-gcc-switches -flto=4 -Werror=odr -Werror=lto-type-mismatch -Werror=strict-aliasing  -Wformat -Werror=format-security -DNDEBUG -msse -mfpmath=sse -msse2 -std=gnu++17   -fschedule-insns2 -fomit-frame-pointer -Wreturn-type -fno-math-errno -fno-signaling-nans -fsigned-zeros -fno-associative-math -pthread -fvisibility=hidden -fPIC -MD -MT lang/CMakeFiles/libsclang.dir/LangSource/PyrParseNode.cpp.o -MF lang/CMakeFiles/libsclang.dir/LangSource/PyrParseNode.cpp.o.d -o lang/CMakeFiles/libsclang.dir/LangSource/PyrParseNode.cpp.o -c /var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource/PyrParseNode.cpp
FAILED: lang/CMakeFiles/libsclang.dir/LangSource/PyrParseNode.cpp.o 
/usr/bin/x86_64-pc-linux-gnu-g++ -DBOOST_CHRONO_HEADER_ONLY -DBOOST_CONFIG_SUPPRESS_OUTDATED_MESSAGE -DHAVE_ALSA=1 -DHAVE_HIDAPI -DHAVE_LID -DHAVE_READLINE -DNO_INTERNAL_SERVER -DQT_COLLIDER_EXPORTING -DQT_CORE_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_NO_DEBUG -DQT_NO_KEYWORDS -DQT_PRINTSUPPORT_LIB -DQT_SVG_LIB -DQT_WIDGETS_LIB -DSC_AUDIO_API_PORTAUDIO -DSC_DATA_DIR=\"/usr/share/SuperCollider\" -DSC_HIDAPI -DSC_IDE -DSC_QT -DSNDFILE_HAS_MPEG -DSNDFILE_HAS_OPUS -DSNDFILE_HAS_VORBIS -DYYSTACK_USE_ALLOC -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999_build/common -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/include/common -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/include/lang -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/include/plugin_interface -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/include/server -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/common -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/external_libraries/boost_sync/include -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangPrimSource -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/external_libraries/nova-tt -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource/Bison -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/external_libraries/hidapi/hidapi -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/external_libraries/hidapi/hidapi_parser -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/QtCollider -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/X11_INCLUDE_DIR -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/SCDoc -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/external_libraries/TLSF-2.4.6/src -isystem /usr/include/qt5 -isystem /usr/include/qt5/QtCore -isystem /usr/lib64/qt5/mkspecs/linux-g++ -isystem /usr/include/qt5/QtGui -isystem /usr/include/qt5/QtNetwork -isystem /usr/include/qt5/QtPrintSupport -isystem /usr/include/qt5/QtWidgets -isystem /usr/include/qt5/QtSvg  -march=native -fstack-protector-all -O2 -pipe -fdiagnostics-color=always -frecord-gcc-switches -flto=4 -Werror=odr -Werror=lto-type-mismatch -Werror=strict-aliasing  -Wformat -Werror=format-security -DNDEBUG -msse -mfpmath=sse -msse2 -std=gnu++17   -fschedule-insns2 -fomit-frame-pointer -Wreturn-type -fno-math-errno -fno-signaling-nans -fsigned-zeros -fno-associative-math -pthread -fvisibility=hidden -fPIC -MD -MT lang/CMakeFiles/libsclang.dir/LangSource/PyrParseNode.cpp.o -MF lang/CMakeFiles/libsclang.dir/LangSource/PyrParseNode.cpp.o.d -o lang/CMakeFiles/libsclang.dir/LangSource/PyrParseNode.cpp.o -c /var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource/PyrParseNode.cpp
In file included from /var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource/PyrParseNode.h:24,
                 from /var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource/PyrParseNode.cpp:22:
/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource/PyrParseNode.cpp: In member function ‘virtual void PyrCallNodeBase::compilePartialApplication(int, PyrSlot*)’:
/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource/PyrKernel.h:144:23: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
  144 | #define METHRAW(obj) ((PyrMethodRaw*)&(((PyrBlock*)obj)->rawData1))
      |                      ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource/PyrParseNode.cpp:1792:9: note: in expansion of macro ‘METHRAW’
 1792 |         METHRAW(prevBlock)->needsHeapContext = 1;
      |         ^~~~~~~
/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource/PyrParseNode.cpp: In member function ‘virtual void PyrCallNode::compileCall(PyrSlot*)’:
/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource/PyrKernel.h:144:23: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
  144 | #define METHRAW(obj) ((PyrMethodRaw*)&(((PyrBlock*)obj)->rawData1))
      |                      ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource/PyrParseNode.cpp:1835:24: note: in expansion of macro ‘METHRAW’
 1835 |     int numBlockArgs = METHRAW(gCompilingBlock)->numargs;
      |                        ^~~~~~~
/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource/PyrParseNode.cpp: In member function ‘virtual void PyrSlotNode::compilePushLit(PyrSlot*)’:
/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource/PyrKernel.h:144:23: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
  144 | #define METHRAW(obj) ((PyrMethodRaw*)&(((PyrBlock*)obj)->rawData1))
      |                      ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource/PyrParseNode.cpp:3152:17: note: in expansion of macro ‘METHRAW’
 3152 |                 METHRAW(gCompilingBlock)->needsHeapContext = 1;
      |                 ^~~~~~~
/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource/PyrParseNode.cpp: In member function ‘virtual void PyrSlotNode::compileLiteral(PyrSlot*)’:
/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource/PyrKernel.h:144:23: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
  144 | #define METHRAW(obj) ((PyrMethodRaw*)&(((PyrBlock*)obj)->rawData1))
      |                      ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource/PyrParseNode.cpp:3222:17: note: in expansion of macro ‘METHRAW’
 3222 |                 METHRAW(gCompilingBlock)->needsHeapContext = 1;
      |                 ^~~~~~~
cc1plus: some warnings being treated as errors
[3/6] /usr/bin/x86_64-pc-linux-gnu-g++ -DBOOST_CHRONO_HEADER_ONLY -DBOOST_CONFIG_SUPPRESS_OUTDATED_MESSAGE -DHAVE_ALSA=1 -DHAVE_HIDAPI -DHAVE_LID -DHAVE_READLINE -DNO_INTERNAL_SERVER -DQT_COLLIDER_EXPORTING -DQT_CORE_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_NO_DEBUG -DQT_NO_KEYWORDS -DQT_PRINTSUPPORT_LIB -DQT_SVG_LIB -DQT_WIDGETS_LIB -DSC_AUDIO_API_PORTAUDIO -DSC_DATA_DIR=\"/usr/share/SuperCollider\" -DSC_HIDAPI -DSC_IDE -DSC_QT -DSNDFILE_HAS_MPEG -DSNDFILE_HAS_OPUS -DSNDFILE_HAS_VORBIS -DYYSTACK_USE_ALLOC -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999_build/common -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/include/common -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/include/lang -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/include/plugin_interface -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/include/server -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/common -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/external_libraries/boost_sync/include -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangPrimSource -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/external_libraries/nova-tt -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource/Bison -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/external_libraries/hidapi/hidapi -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/external_libraries/hidapi/hidapi_parser -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/QtCollider -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/X11_INCLUDE_DIR -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/SCDoc -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/external_libraries/TLSF-2.4.6/src -isystem /usr/include/qt5 -isystem /usr/include/qt5/QtCore -isystem /usr/lib64/qt5/mkspecs/linux-g++ -isystem /usr/include/qt5/QtGui -isystem /usr/include/qt5/QtNetwork -isystem /usr/include/qt5/QtPrintSupport -isystem /usr/include/qt5/QtWidgets -isystem /usr/include/qt5/QtSvg  -march=native -fstack-protector-all -O2 -pipe -fdiagnostics-color=always -frecord-gcc-switches -flto=4 -Werror=odr -Werror=lto-type-mismatch -Werror=strict-aliasing  -Wformat -Werror=format-security -DNDEBUG -msse -mfpmath=sse -msse2 -std=gnu++17   -fschedule-insns2 -fomit-frame-pointer -Wreturn-type -fno-math-errno -fno-signaling-nans -fsigned-zeros -fno-associative-math -pthread -fvisibility=hidden -fPIC -MD -MT lang/CMakeFiles/libsclang.dir/LangSource/PyrInterpreter3.cpp.o -MF lang/CMakeFiles/libsclang.dir/LangSource/PyrInterpreter3.cpp.o.d -o lang/CMakeFiles/libsclang.dir/LangSource/PyrInterpreter3.cpp.o -c /var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource/PyrInterpreter3.cpp
FAILED: lang/CMakeFiles/libsclang.dir/LangSource/PyrInterpreter3.cpp.o 
/usr/bin/x86_64-pc-linux-gnu-g++ -DBOOST_CHRONO_HEADER_ONLY -DBOOST_CONFIG_SUPPRESS_OUTDATED_MESSAGE -DHAVE_ALSA=1 -DHAVE_HIDAPI -DHAVE_LID -DHAVE_READLINE -DNO_INTERNAL_SERVER -DQT_COLLIDER_EXPORTING -DQT_CORE_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_NO_DEBUG -DQT_NO_KEYWORDS -DQT_PRINTSUPPORT_LIB -DQT_SVG_LIB -DQT_WIDGETS_LIB -DSC_AUDIO_API_PORTAUDIO -DSC_DATA_DIR=\"/usr/share/SuperCollider\" -DSC_HIDAPI -DSC_IDE -DSC_QT -DSNDFILE_HAS_MPEG -DSNDFILE_HAS_OPUS -DSNDFILE_HAS_VORBIS -DYYSTACK_USE_ALLOC -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999_build/common -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/include/common -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/include/lang -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/include/plugin_interface -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/include/server -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/common -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/external_libraries/boost_sync/include -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangPrimSource -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/external_libraries/nova-tt -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource/Bison -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/external_libraries/hidapi/hidapi -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/external_libraries/hidapi/hidapi_parser -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/QtCollider -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/X11_INCLUDE_DIR -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/SCDoc -I/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/external_libraries/TLSF-2.4.6/src -isystem /usr/include/qt5 -isystem /usr/include/qt5/QtCore -isystem /usr/lib64/qt5/mkspecs/linux-g++ -isystem /usr/include/qt5/QtGui -isystem /usr/include/qt5/QtNetwork -isystem /usr/include/qt5/QtPrintSupport -isystem /usr/include/qt5/QtWidgets -isystem /usr/include/qt5/QtSvg  -march=native -fstack-protector-all -O2 -pipe -fdiagnostics-color=always -frecord-gcc-switches -flto=4 -Werror=odr -Werror=lto-type-mismatch -Werror=strict-aliasing  -Wformat -Werror=format-security -DNDEBUG -msse -mfpmath=sse -msse2 -std=gnu++17   -fschedule-insns2 -fomit-frame-pointer -Wreturn-type -fno-math-errno -fno-signaling-nans -fsigned-zeros -fno-associative-math -pthread -fvisibility=hidden -fPIC -MD -MT lang/CMakeFiles/libsclang.dir/LangSource/PyrInterpreter3.cpp.o -MF lang/CMakeFiles/libsclang.dir/LangSource/PyrInterpreter3.cpp.o.d -o lang/CMakeFiles/libsclang.dir/LangSource/PyrInterpreter3.cpp.o -c /var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource/PyrInterpreter3.cpp
In file included from /var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource/PyrInterpreter3.cpp:28:
/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource/PyrInterpreter3.cpp: In function ‘void Interpret(VMGlobals*)’:
/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource/PyrKernel.h:144:23: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
  144 | #define METHRAW(obj) ((PyrMethodRaw*)&(((PyrBlock*)obj)->rawData1))
      |                      ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource/PyrInterpreter3.cpp:1595:29: note: in expansion of macro ‘METHRAW’
 1595 |             numArgsPushed = METHRAW(g->block)->numargs;
      |                             ^~~~~~~
/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource/PyrKernel.h:144:23: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
  144 | #define METHRAW(obj) ((PyrMethodRaw*)&(((PyrBlock*)obj)->rawData1))
      |                      ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource/PyrInterpreter3.cpp:1609:29: note: in expansion of macro ‘METHRAW’
 1609 |             numArgsPushed = METHRAW(g->block)->numargs;
      |                             ^~~~~~~
/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource/PyrKernel.h:144:23: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
  144 | #define METHRAW(obj) ((PyrMethodRaw*)&(((PyrBlock*)obj)->rawData1))
      |                      ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource/PyrInterpreter3.cpp:1624:29: note: in expansion of macro ‘METHRAW’
 1624 |             numArgsPushed = METHRAW(g->block)->numargs;
      |                             ^~~~~~~
/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource/PyrKernel.h:144:23: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
  144 | #define METHRAW(obj) ((PyrMethodRaw*)&(((PyrBlock*)obj)->rawData1))
      |                      ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource/PyrInterpreter3.cpp:1639:29: note: in expansion of macro ‘METHRAW’
 1639 |             numArgsPushed = METHRAW(g->block)->numargs;
      |                             ^~~~~~~
/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource/PyrKernel.h:144:23: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
  144 | #define METHRAW(obj) ((PyrMethodRaw*)&(((PyrBlock*)obj)->rawData1))
      |                      ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource/PyrInterpreter3.cpp:1654:29: note: in expansion of macro ‘METHRAW’
 1654 |             numArgsPushed = METHRAW(g->block)->numargs + 1;
      |                             ^~~~~~~
/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource/PyrKernel.h:144:23: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
  144 | #define METHRAW(obj) ((PyrMethodRaw*)&(((PyrBlock*)obj)->rawData1))
      |                      ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/var/tmp/portage/media-sound/supercollider-9999/work/supercollider-9999/lang/LangSource/PyrInterpreter3.cpp:1669:29: note: in expansion of macro ‘METHRAW’
 1669 |             numArgsPushed = METHRAW(g->block)->numargs + 1;
      |                             ^~~~~~~
cc1plus: some warnings being treated as errors
ninja: build stopped: cannot make progress due to previous errors.
```

Downstream report: https://bugs.gentoo.org/927071
Full build log: [build.log](https://github.com/supercollider/supercollider/files/14832452/build.log)


Note the downstream report is reported against 3.13.0 but my local reproducer is against git develop.

--- a/HelpSource/Classes/GVerb.schelp
+++ b/HelpSource/Classes/GVerb.schelp
@@ -9,8 +9,7 @@ A two-channel reverb link::Classes/UGen::, based on the "GVerb" LADSPA effect by
 subsection:: Known issues
 list::
 ## There is a large CPU spike when the synth is instantiated while all the delay lines are zeroed out.
-## Quick changes in roomsize result in zipper noise.
-## emphasis::Changing the roomsize does not work properly! Still trying to look for the bug... (-josh)::
+## Changes in roomsize result in pitch shifting and noise.
 ::
 
 classmethods::

--- a/include/common/SC_Types.h
+++ b/include/common/SC_Types.h
@@ -59,6 +59,7 @@ typedef union {
 const unsigned int kSCNameLen = 8;
 const unsigned int kSCNameByteLen = 8 * sizeof(int32);
 
+// Do not use this. C casting is bad and causes many subtle issues.
 #ifdef __GXX_EXPERIMENTAL_CXX0X__
 #    define sc_typeof_cast(x) (decltype(x))
 #elif defined(__GNUC__)

--- a/include/plugin_interface/SC_Unit.h
+++ b/include/plugin_interface/SC_Unit.h
@@ -21,6 +21,7 @@
 
 #pragma once
 
+#include <type_traits>
 #include "SC_Types.h"
 #include "SC_SndBuf.h"
 
@@ -91,8 +92,13 @@ enum { kUnitDef_CantAliasInputsToOutputs = 1 };
         ClearUnitOnMemFailed                                                                                           \
     }
 
+template <typename ToType, typename Value>
+[[nodiscard]] inline constexpr auto copyAndCastToTypeOfFirstArg(const ToType&, const Value& value) noexcept {
+    using TargetT = std::remove_cv_t<std::remove_reference_t<ToType>>;
+    return static_cast<TargetT>(value);
+}
 // calculate a slope for control rate interpolation to audio rate.
-#define CALCSLOPE(next, prev) ((next - prev) * sc_typeof_cast(next) unit->mRate->mSlopeFactor)
+#define CALCSLOPE(next, prev) ((next - prev) * copyAndCastToTypeOfFirstArg(next, unit->mRate->mSlopeFactor))
 
 // get useful values
 #define SAMPLERATE (unit->mRate->mSampleRate)

--- a/server/plugins/ReverbUGens.cpp
+++ b/server/plugins/ReverbUGens.cpp
@@ -1063,7 +1063,7 @@ int nearestprime(int n, float rerror) {
     if (isprime(n))
         return (n);
     /* assume n is large enough and n*rerror enough smaller than n */
-    bound = (int)(n * rerror);
+    bound = n * static_cast<int>(rerror);
     for (k = 1; k <= bound; k++) {
         if (isprime(n + k))
             return (n + k);
@@ -1162,7 +1162,7 @@ static inline void damper_set(GVerb* unit, g_damper* p, float damping) { p->damp
 
 static inline float damper_do(GVerb* unit, g_damper* p, float x) {
     float y;
-    y = x * (1.0 - p->damping) + p->delay * p->damping;
+    y = x * (1.0f - p->damping) + p->delay * p->damping;
     p->delay = zapgremlins(y);
     return (y);
 }
@@ -1182,7 +1182,7 @@ static inline void gverb_set_roomsize(GVerb* unit, const float a) {
         unit->roomsize = 1.0;
     } else {
         if (a >= unit->maxroomsize)
-            unit->roomsize = unit->maxroomsize - 1.;
+            unit->roomsize = unit->maxroomsize - 1.f;
         else
             unit->roomsize = a;
     };
@@ -1197,7 +1197,7 @@ static inline void gverb_set_roomsize(GVerb* unit, const float a) {
 
     for (i = 0; i < FDNORDER; i++) {
         float oldfdngain = unit->fdngains[i];
-        unit->fdngains[i] = -powf((float)unit->alpha, unit->fdnlens[i]);
+        unit->fdngains[i] = static_cast<float>(-std::pow(unit->alpha, static_cast<double>(unit->fdnlens[i])));
         unit->fdngainslopes[i] = CALCSLOPE(unit->fdngains[i], oldfdngain);
     }
 
@@ -1208,7 +1208,7 @@ static inline void gverb_set_roomsize(GVerb* unit, const float a) {
 
     for (i = 0; i < FDNORDER; i++) {
         float oldtapgain = unit->tapgains[i];
-        unit->tapgains[i] = pow(unit->alpha, unit->taps[i]);
+        unit->tapgains[i] = static_cast<float>(std::pow(unit->alpha, static_cast<double>(unit->taps[i])));
         unit->tapgainslopes[i] = CALCSLOPE(unit->tapgains[i], oldtapgain);
     }
 }
@@ -1226,7 +1226,7 @@ static inline void gverb_set_revtime(GVerb* unit, float a) {
 
     for (i = 0; i < FDNORDER; i++) {
         float oldfdngain = unit->fdngains[i];
-        unit->fdngains[i] = -powf((float)unit->alpha, unit->fdnlens[i]);
+        unit->fdngains[i] = static_cast<float>(-std::pow(unit->alpha, static_cast<double>(unit->fdnlens[i])));
         unit->fdngainslopes[i] = CALCSLOPE(unit->fdngains[i], oldfdngain);
     }
 }
@@ -1277,7 +1277,9 @@ void GVerb_Ctor(GVerb* unit) {
     unit->earlylevel = 0.; // IN0(7);
     unit->taillevel = 0.; // IN0(8);
 
-    float maxroomsize = unit->maxroomsize = IN0(9);
+    // when roomsize is greater than maxroomsize, it is set to maxroomsize - 1
+    // when roomsize is less than 0, it is set to 1, therefore maxroomsize must be at least 1.
+    float maxroomsize = unit->maxroomsize = sc_max(1.0001f, IN0(9));
 
     float maxdelay = unit->maxdelay = SAMPLERATE * maxroomsize / 340.f;
     float largestdelay = unit->largestdelay = SAMPLERATE * roomsize / 340.f;
@@ -1305,7 +1307,7 @@ void GVerb_Ctor(GVerb* unit) {
         } else {
             unit->fdnlens[i] = f_round(gb);
         }
-        unit->fdngains[i] = -powf((float)alpha, unit->fdnlens[i]);
+        unit->fdngains[i] = static_cast<float>(-std::pow(alpha, static_cast<double>(unit->fdnlens[i])));
     }
     // make the fixeddelay lines and dampers
     for (int i = 0; i < FDNORDER; i++) {
@@ -1315,7 +1317,7 @@ void GVerb_Ctor(GVerb* unit) {
     }
 
     // diffuser section
-    float diffscale = (float)unit->fdnlens[3] / (210. + 159. + 562. + 410.);
+    float diffscale = static_cast<float>(unit->fdnlens[3] / (210. + 159. + 562. + 410.));
     float spread1 = spread;
     float spread2 = 3.0 * spread;
 
@@ -1359,7 +1361,7 @@ void GVerb_Ctor(GVerb* unit) {
     unit->taps[3] = 5; //+ f_round(0.000 * largestdelay);
 
     for (int i = 0; i < FDNORDER; i++) {
-        unit->tapgains[i] = pow(alpha, (double)unit->taps[i]);
+        unit->tapgains[i] = static_cast<float>(std::pow(alpha, static_cast<double>(unit->taps[i])));
     }
 
     unit->tapdelay = make_fixeddelay(unit, 44000, 44000);

